### ISSUE TITLE
Null Object Reference (when completing a battle)

### ISSUE THREAD

**Issue by nospoone** [larsiusprime/tdrpg-bugs#665]
# Basic Info
```
OS: Windows 10 (TH2)
Version: 0.4.5-1-g7f78dd9 - demo, build 1
```
**Windowed** at `1600x1200`

# Steps to Reproduce:
Not sure, but here goes~

1. Start a new game
2. Start a new battle
3. Have a unit die right before the battle ends

I might have tried to place a new one as the battle ended. It happened very quickly!

# Happened
As the battle ended, I got a Null Object Reference error, and the game closed.

![image](https://cloud.githubusercontent.com/assets/708588/14322306/53a42344-fbea-11e5-9fb2-564924b4f45e.png)

# Expected
Not getting an error :smile: 

---

_This was probably caught by Crashdumper - it happened around 11:20EST on April 6th, 2016_

**Comment by sruloart** [comment]
It also happens when you let Azra be slain very quickly (at X8 or X16 the normal speed).

**Comment by larsiusprime** [comment]
Having a real hard time reproducing this... man I wish crashdumper was working with OpenFL-Next, but I don't think it is. Will have to fix that.

**Comment by ManDude** [comment]
I think I had this problem before and it was related to speed, but I didn't look much into it. I'll see if I can reproduce this too.

**Comment by ezacharyk** [comment]
I am not sure what may have caused this for me. I was in the second battle in Advanced mode. I lost the battle and the game crashed. I had lost both my berserkers just before losing as well. I don't think it happened at the same time, but it happened shortly before. I played it a second time and the game continued just fine. However, there are some rather severe frame rate drops during this battle.

**Comment by larsiusprime** [comment]
Were you using the debug or the release build?

On Fri, Apr 8, 2016 at 5:05 PM, E. Zachary Knight <notifications@github.com>
wrote:

> I am not sure what may have caused this for me. I was in the second battle
> in Advanced mode. I lost the battle and the game crashed. I had lost both
> my berserkers just before losing as well. I don't think it happened at the
> same time, but it happened shortly before. I played it a second time and
> the game continued just fine. However, there are some rather severe frame
> rate drops during this battle.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/larsiusprime/tdrpg-bugs/issues/665#issuecomment-207625041>
>



-- 
www.fortressofdoors.com -- Games, Art, Design

**Comment by ezacharyk** [comment]
The debug version.

**Comment by larsiusprime** [comment]
That'll be why. If you run the debug version from a command line, it will
sometimes spit out a stack trace on exit, but it has much worse
performance. The release version runs better but provides less useful
information.

On Fri, Apr 8, 2016 at 5:09 PM, E. Zachary Knight <notifications@github.com>
wrote:

> The debug version.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/larsiusprime/tdrpg-bugs/issues/665#issuecomment-207625858>
>



-- 
www.fortressofdoors.com -- Games, Art, Design

**Comment by ezacharyk** [comment]
Unfortunately I forgot to run it from the command line the one time it crashed.

**Comment by larsiusprime** [comment]
My suspicion is that it's caused by losing a defender -- they start dying the moment their hp is reduced to 0, but they don't actually "finish" dying until a bit later and are then removed from the board. I bet it's related to them dying with exactly perfect timing, that interacts with the battle ending, a "remove defender" signal being triggered, and some expected object is null as the battle is in the process of shutting down.

Man I really need to get crashdumper working again, that would catch this immediately.

**Comment by nospoone** [comment]
This seems like exactly what happened for me.

**Comment by larsiusprime** [comment]
Oh btw @nospoone, @ezacharyk, can you export your save files and post them for me to download? Maybe there's something specific to your setups that makes it easier to reproduce.

**Comment by larsiusprime** [comment]
Nevermind, reproduced it!

I set up a battle with one enemy left, summoned twelve defenders, killed them (with a cheat key) one by one with a tiny time offset in between, then killed the last enemy. Sure enough, one of them triggered the null error!

**Comment by larsiusprime** [comment]
I'm tentatively closing this as fixed as after a patch I can't reproduce the error anymore. It might have been only a similar bug I caught though, so feel free to re-open if it recurs again post version 0.4.7.

**Comment by IBwWG** [comment]
Where do you get 0.4.7?  I only found 0.4.5 HD demo on your blog, and with it ran into this crash.

**Comment by IBwWG** [comment]
In fact, I just had it happen also without my defenders dying, nor it being the last enemy.  I wonder if it would happen on 0.4.7...

![dq1hd-crash](https://cloud.githubusercontent.com/assets/16719964/14637720/7131a8a2-0633-11e6-9452-c426bdf48152.png)

Are you getting the crash info from it automatically again nowadays, or should I send more info?  This is on Win7x64.

**Comment by ManDude** [comment]
That level will ALWAYS give a Null Object Reference error if you play it in Extreme difficulty. I think I had reported that one already? Hmm.

**Comment by IBwWG** [comment]
@ManDude Not always, as I said the second time through it was OK.

**Comment by IBwWG** [comment]
Had it happen on the last level just now, no healers (so not #679) and nobody else died immediately beforehand (although one of my defenders died early on) but when Azra died I got a null ref:

![dq1hd-crash8](https://cloud.githubusercontent.com/assets/16719964/14645198/05492bac-0655-11e6-86ad-9f3bc62be9d1.png)

For the record, I had Azra die on another level (I believe on Extreme, but definitely on 16X) without a crash.  Win7x64, DQ 0.4.5 windowed just below 1920x1080 with that desktop res in multimon env.

**Comment by larsiusprime** [comment]
Thanks @IBwWG ... so this folder has all the latest stuff:
https://drive.google.com/drive/folders/0B82nNNq_hucHOEV1eF9pQ1B1eXM

I need to just update the blog to point to that directly I guess.

**Comment by IBwWG** [comment]
Awesome, I'll give it a go.  I just had the healer bug again on Extreme on this level, but a) it didn't happen on the very first heal...maybe the 6th or 7th and b) maybe it won't happen anymore once I update.  :)

**Comment by larsiusprime** [comment]
Okay I reproduced it! Debugging now...

**Comment by larsiusprime** [comment]
Okay the crash seems to have nothing to do with your save files and is just a generic error with the demo build.

**Comment by larsiusprime** [comment]
Okay think I've gotten it narrowed down, I think the way I'm preprocessing the demo build is creating an invalid xml file.

**Comment by IBwWG** [comment]
Once 0.5.2 is out I'll give another run from the beginning and see if I catch this again.

BTW, if you buy DQ1, do you get access to the full HD builds at this point?

**Comment by larsiusprime** [comment]
The idea is that all DQ1 owners will receive a free upgrade to DQHD. It would be cool if every current DQ1 owner had access to the shared folder but I don't have any logistically sane way to set that up for 250,000+ owners.

If you want I can give you access to the full builds, as our policy is to make that open to anyone who dedicates decent test reports.

**Comment by farmboy0** [comment]
Just had this crash when I didnt place a defender and let the battle end on 16x
Demo version: (0.5.1 - demo, build 1) (Linux 64-bit)

Called from ApplicationMain::main ApplicationMain.hx line 159
Called from ApplicationMain::create ApplicationMain.hx line 71
Called from lime.app.Application::exec lime/app/Application.hx line 222
Called from lime._backend.native.NativeApplication::exec lime/_backend/native/NativeApplication.hx line 131
Called from lime._backend.native.NativeApplication::handleRenderEvent lime/_backend/native/NativeApplication.hx line 363
Called from lime.app.Event_Void_Void::dispatch lime/app/Event.hx line 145
Called from *::_Function_2_1 lime/app/Application.hx line 110
Called from lime.app.Application::render lime/app/Application.hx line 762
Called from openfl.display.Stage::render openfl/display/Stage.hx line 677
Called from openfl.display.DisplayObjectContainer::__broadcast openfl/display/DisplayObjectContainer.hx line 355
Called from openfl.display.DisplayObject::__broadcast openfl/display/DisplayObject.hx line 251
Called from openfl.events.EventDispatcher::__dispatchEvent openfl/events/EventDispatcher.hx line 237
Called from flixel.FlxGame::onEnterFrame flixel/FlxGame.hx line 529
Called from flixel.FlxGame::step flixel/FlxGame.hx line 662
Called from com.leveluplabs.tdrpg.Main::update com/leveluplabs/tdrpg/Main.hx line 1462
Called from flixel.FlxGame::update flixel/FlxGame.hx line 751
Called from flixel.FlxState::tryUpdate flixel/FlxState.hx line 168
Called from com.leveluplabs.tdrpg.State_Battle::update com/leveluplabs/tdrpg/State_Battle.hx line 868
Called from com.leveluplabs.tdrpg.State_Battle::battleUpdate com/leveluplabs/tdrpg/State_Battle.hx line 983
Called from com.leveluplabs.tdrpg.BattleModel::update com/leveluplabs/tdrpg/BattleModel.hx line 1395
Called from com.leveluplabs.tdrpg.BattleModel::doUpdate com/leveluplabs/tdrpg/BattleModel.hx line 1797
Called from com.leveluplabs.tdrpg.BattleModel::updateEnemies com/leveluplabs/tdrpg/BattleModel.hx line 3174
Called from com.leveluplabs.tdrpg.EnemySprite::updateFromModel com/leveluplabs/tdrpg/EnemySprite.hx line 510
Called from com.leveluplabs.tdrpg.EnemySprite::updateWalk com/leveluplabs/tdrpg/EnemySprite.hx line 2150
Called from com.leveluplabs.tdrpg.EnemySprite::getNextWalk com/leveluplabs/tdrpg/EnemySprite.hx line 1115
Called from com.leveluplabs.tdrpg.BattleModel::getNextPathTile com/leveluplabs/tdrpg/BattleModel.hx line 1954
Called from com.leveluplabs.tdrpg.BattleModel::getWalkMap com/leveluplabs/tdrpg/BattleModel.hx line 2038
Called from com.leveluplabs.tdrpg.WalkMapIndex::get com/leveluplabs/tdrpg/WalkMapIndex.hx line 71
Called from flixel.tile.FlxBaseTilemap::loadMapFromArray flixel/tile/FlxBaseTilemap.hx line 256
Error : Null Object Reference

**Comment by IBwWG** [comment]
Just tried the ozimal coat battle on 0.5.2 demo, win7x64, and had it crash at the end (when Azra died):

```
FlxUIButton name = movies_btn
FlxUIButton name = music_btn
FlxUIButton name = sfx_btn
FlxUIButton name = party_btn
FlxUIButton name = btn_journal
FlxUIButton name = btn_town
FlxUIButton name = tips_btn
FlxUIButton name = btn_battle
FlxUIButton name = exit_btn
FlxUIButton name = options_btn
FlxUIButton name = menu_btn
FlxUIButton name = bonus_btn
Called from ApplicationMain::main ApplicationMain.hx line 159
Called from ApplicationMain::create ApplicationMain.hx line 71
Called from lime.app.Application::exec lime/app/Application.hx line 222
Called from lime._backend.native.NativeApplication::exec lime/_backend/native/Na
tiveApplication.hx line 136
Called from lime._backend.native.NativeApplication::handleRenderEvent lime/_back
end/native/NativeApplication.hx line 368
Called from lime.app.Event_Void_Void::dispatch lime/app/Event.hx line 145
Called from *::_Function_2_1 lime/app/Application.hx line 110
Called from lime.app.Application::render lime/app/Application.hx line 762
Called from openfl.display.Stage::render openfl/display/Stage.hx line 677
Called from openfl.display.DisplayObjectContainer::__broadcast openfl/display/Di
splayObjectContainer.hx line 355
Called from openfl.display.DisplayObject::__broadcast openfl/display/DisplayObje
ct.hx line 251
Called from openfl.events.EventDispatcher::__dispatchEvent openfl/events/EventDi
spatcher.hx line 237
Called from flixel.FlxGame::onEnterFrame flixel/FlxGame.hx line 529
Called from flixel.FlxGame::step flixel/FlxGame.hx line 662
Called from com.leveluplabs.tdrpg.Main::update com/leveluplabs/tdrpg/Main.hx lin
e 1462
Called from flixel.FlxGame::update flixel/FlxGame.hx line 751
Called from flixel.FlxState::tryUpdate flixel/FlxState.hx line 168
Called from com.leveluplabs.tdrpg.State_Battle::update com/leveluplabs/tdrpg/Sta
te_Battle.hx line 868
Called from com.leveluplabs.tdrpg.State_Battle::battleUpdate com/leveluplabs/tdr
pg/State_Battle.hx line 983
Called from com.leveluplabs.tdrpg.BattleModel::update com/leveluplabs/tdrpg/Batt
leModel.hx line 1395
Called from com.leveluplabs.tdrpg.BattleModel::doUpdate com/leveluplabs/tdrpg/Ba
ttleModel.hx line 1797
Called from com.leveluplabs.tdrpg.BattleModel::updateEnemies com/leveluplabs/tdr
pg/BattleModel.hx line 3174
Called from com.leveluplabs.tdrpg.EnemySprite::updateFromModel com/leveluplabs/t
drpg/EnemySprite.hx line 510
Called from com.leveluplabs.tdrpg.EnemySprite::updateWalk com/leveluplabs/tdrpg/
EnemySprite.hx line 2150
Called from com.leveluplabs.tdrpg.EnemySprite::getNextWalk com/leveluplabs/tdrpg
/EnemySprite.hx line 1115
Called from com.leveluplabs.tdrpg.BattleModel::getNextPathTile com/leveluplabs/t
drpg/BattleModel.hx line 1954
Called from com.leveluplabs.tdrpg.BattleModel::getWalkMap com/leveluplabs/tdrpg/
BattleModel.hx line 2038
Called from com.leveluplabs.tdrpg.WalkMapIndex::get com/leveluplabs/tdrpg/WalkMa
pIndex.hx line 71
Called from flixel.tile.FlxBaseTilemap::loadMapFromArray flixel/tile/FlxBaseTile
map.hx line 256
```

**Comment by larsiusprime** [comment]
Oh shoot this is a subtle one -- an enemy is hitting its next tile and asking for the next location at the exact second the entire battle data becomes invalid!

**Comment by IBwWG** [comment]
Whew!

**Comment by larsiusprime** [comment]
Alrighty, the best I can really do here is throw in a bunch of null checks around things and hope that catches it. I'm 95% sure this will work.

Because the timing is so hard to reproduce here, what I'm gonna do is assume this is fixed, and we'll re-open it if we see this same stack trace signature again.

This is (presumably) fixed in 0.5.4.
